docs/proposals/hosted_components_templates.md

```diff
@@ -0,0 +1,564 @@
+# Use [OCP templates and] APBs to install hosted components
+
+## Description
+Currently most of our hosted components have their own repositories where they
+maintain changes and then need to synchronize those changes within their
+respective roles. This can cause delays and times where features don't
+match up with what is installed or can be configured.
+
+There is also a concern that copying files from one location to another can
+cause cases where files are out of sync or versions don't align with what is
+installed and what is configured (is what we configured correct for the image
+version we installed).
+
+## Rationale
+As was piloted with the Template Service Broker ("TSB" for short), a role was
+created to process and configure OCP templates that were synchronized (currently
+still a manual process) from the origin repo where the component resides.
+
+This allowed any changes for the installation to be maintained by the developers
+and keep them from needing to be fluent with Ansible. It also afforded a level
+of abstraction from needing to keep in sync with TSB related changes and then
+translate them to changes within an Ansible role.
+
+However, there is still the gap of ensuring a role configures changes appropriately
+for the version of the component it is installing and we are violating the "DRY"
+principle by needing to sync files between repos.
+
+Prior to 3.4, Aggregated Logging solved this using a deployer pod. Everything
```

Comment:
I think my concerns with this current approach bubble up from this prior time.  I was on the operations team (prior to 3.4) when the deployer pod would come up, run a shell script, go away and the software was either configured or it wasn't. As a system administrator this made config management a no-go when deploying with this pod.  How were changes updated, placed inside the logging container, managed?  The deployer pod would generate keys and certificates that were then only known to the logging system.   After all of this when we attempted to go from release to release there was no upgrade process.  The process was to remove the current and redeploy which also if my memory servers would remove the pvc's and then remove the pv's which would remove the current logs.

Having stated the concerns from the previous approach, I do like that the logging team wants to own their code and not have to go between their own repository and openshift-ansible.  Delivering the hosted components from a separate channel sounds nice as well since there are few that understand the inner workings of some of the complex hosted components, who better to manage them than the team that created them.

Let's start with a few questions:
How does one idempotently update values of the currently deployed logging components?
How does one config manage the settings in this new process?
How do you envision the release process?

Reply:
> How does one idempotently update values of the currently deployed logging components?

IIRC we had an upgrade option that one could specify for the deployer pod that would simply `oc patch` existing API objects which was rather fragile but with our current implementation (the logging roles) we leverage modules that `oc apply` to 'make it so'. 

From a standpoint of support, I feel that the latter is much preferred and I think that is what we would continue to do since we would effectively port the `openshift_logging*` roles into a playbook bundle. 

I'm not sure if it would make sense to have a 'validation' entrypoint that one could run afterwards to verify that everything ran to completion and correctly. @fabianvf are there any discussions around that currently?

> How does one config manage the settings in this new process?

I may not be understanding your question here correctly, but it would make sense that we still provide configuration values as inventory variables or as values on the command line that would be passed on to the APB for configuring the component. From an Ansible [role] user standpoint, there should be no change from what is currently done.

> How do you envision the release process?

During a release each team would provide an artifact (their APB) that their Ansible role would be consuming. So that should follow the same cadence as other images being released, meaning the component tags should match the OCP tag/version. Likewise, when a fix is required you would see a release bump for that image.

Reply:
> I may not be understanding your question here correctly, but it would make sense that we still provide configuration values as inventory variables or as values on the command line that would be passed on to the APB for configuring the component. From an Ansible [role] user standpoint, there should be no change from what is currently done.

I don't believe this is the recommended way currently.  If I need to modify settings in config then I make those settings on my cluster either manually, call to openshift-ansible, or any other tools a customer wants to use to manage their cluster.  So that I understand, are you recommending that users rerun the openshift-ansible role to config manage their settings?

In the recent past this might be more feasible as less and less changes are occurring during the openshift-ansible playbook runs.  In the distant past, this would have unexpected changes.  If we can guarantee only the desired changes land then I think this path can be valid.  If there is a high risk of re-configuring, updating, or making unintended changes when rerunning the openshift-ansible playbook to manage logging then this might be too high risk for production.

Reply:
I agree, I think we've been trending towards seeing less and less unintended changes.

To me our roles should "make it so"* and scope them in such a way that we aren't modifying other components.

My thought is that we ultimately want to not need to do any manual customizations after installations with perhaps the exception being capacity specific settings. But even so, it should be possible to deploy a new cluster afterwards with updated inventory values and avoid needing to do the manual changes afterwards.

\* To the best of our ability -- the contents of configmaps can make this tricky if customers are allowed to change values.